src/steps/toolbx.rs

```diff
@@ -0,0 +1,74 @@
+use crate::error::TopgradeError;
+use anyhow::Result;
+
+use crate::executor::ExecutorOutput;
+use crate::terminal::print_separator;
+use crate::{execution_context::ExecutionContext, utils::require};
+use log::debug;
+use std::path::Path;
+use std::{
+    io::{self, Write},
+    process::Command,
+};
+
+fn list_toolboxes(toolbx: &Path) -> Result<Vec<String>> {
+    let output = Command::new(toolbx).args(&["list", "--containers"]).output()?;
+    let output_str = String::from_utf8(output.stdout)?;
+
+    let proc: Vec<String> = output_str
+        .lines()
+        // Skip the first line since that contains only status information
+        .skip(1)
+        .map(|line| match line.split_whitespace().nth(1) {
+            Some(word) => word.to_string(),
+            None => String::from(""),
+        })
+        .filter(|x| !x.is_empty())
+        .collect();
+
+    Ok(proc)
+}
+
+pub fn run_toolbx(ctx: &ExecutionContext) -> Result<()> {
+    print_separator("Toolbx");
+
+    let toolbx = require("toolbox")?;
+    let toolboxes = list_toolboxes(&toolbx)?;
+    debug!("Toolboxes to inspect: {:?}", toolboxes);
+
+    // Path of the running topgrade executable
+    let topgrade_path = String::from("/run/host") + std::env::current_exe()?.as_path().to_str().unwrap();
+
+    for tb in toolboxes.iter() {
+        print!("> Upgrading toolbox '{}'...", tb);
+        // Write out immediately
+        io::stdout().flush()?;
+
+        let output = ctx
+            .run_type()
+            .execute(&toolbx)
+            .args(&["run", "-c", tb, &topgrade_path, "--yes", "--only", "system"])
+            .output()?;
+
+        if let ExecutorOutput::Wet(output) = output {
+            let status = output.status;
+
+            if !status.success() {
+                println!(" FAILED");
+                // Is this ok?
+                io::stderr().write(&output.stderr).ok();
+
+                return Err(TopgradeError::ProcessFailed(status).into());
```

Comment:
This will cancel remaining toolboxes, which is probably unwanted. I suggest that instead of returning you will initialize a boolean variable telling whether we failed or succeeded, and set it whenever a failure in one container occur. At the end of the function an error will be returned if any container failed.

Reply:
I'm afraid I'll need some help with that. Initializing and setting the boolean variable is straightforward, but:

- I don't know what to do then. Do I return some specially crafted `Err`, or how is this done "manually"? Because so far I could only gather from the code it's done implicilty by directly returning what falls out of e.g. `check_run`
- There's a critical error which I currently can't catch: If topgrade fails to run in a toolbox due to a glibc version mismatch, `toolbox run` will not return an error code but 0 instead. This of course gives the impression everything went well, which it didn't. Is there a way to get the stout/stderr of the executor outputs and parse those for the error message?